src/hotspot/share/opto/escape.cpp

```diff
@@ -1223,23 +1223,29 @@ bool ConnectionGraph::reduce_phi_on_safepoints_helper(Node* ophi, Node* cast, No
     nsr_merge_pointer = _igvn->transform(ConstraintCastNode::make_cast_for_type(cast->in(0), cast->in(1), new_t, ConstraintCastNode::RegularDependency, nullptr));
   }
 
+  GrowableArray<Node*> non_debug_edges_worklist;
   for (uint spi = 0; spi < safepoints.size(); spi++) {
     SafePointNode* sfpt = safepoints.at(spi)->as_SafePoint();
     JVMState *jvms      = sfpt->jvms();
     uint merge_idx      = (sfpt->req() - jvms->scloff());
     int debug_start     = jvms->debug_start();
 
+    sfpt->remove_non_debug_edges(non_debug_edges_worklist);
+
     SafePointScalarMergeNode* smerge = new SafePointScalarMergeNode(merge_t, merge_idx);
     smerge->init_req(0, _compile->root());
     _igvn->register_new_node_with_optimizer(smerge);
 
+    assert(sfpt->jvms()->endoff() == sfpt->req(), "no extra edges past debug info allowed");
+
     // The next two inputs are:
     //  (1) A copy of the original pointer to NSR objects.
     //  (2) A selector, used to decide if we need to rematerialize an object
     //      or use the pointer to a NSR object.
     // See more details of these fields in the declaration of SafePointScalarMergeNode
     sfpt->add_req(nsr_merge_pointer);
     sfpt->add_req(selector);
+    sfpt->jvms()->set_endoff(sfpt->req());
```

Comment:
This seems like a subtle change that deserves a comment.  Is it changing the behavior?  Now these two edges are considered part of scalar/debug edges, when before they weren't?

Reply:
> Is it changing the behavior?

It doesn't. The sole purpose of the change is to please the assert in `create_scalarized_object_description()`.

At the end of successful call, `create_scalarized_object_description()` extends debug info to cover all newly added inputs. And the failure is non-recoverable. So, the very first iteration of the loop extends debug info beyond those 2 edges.  

I'll add a comment.